src/main/scala/nl.knaw.dans.easy.deposit/WorkerThread.scala

```diff
@@ -0,0 +1,129 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package nl.knaw.dans.easy.deposit
+
+import java.nio.file.StandardWatchEventKinds.{ ENTRY_CREATE, ENTRY_DELETE }
+import java.nio.file.{ WatchEvent, WatchKey }
+
+import better.files.File
+import nl.knaw.dans.easy.deposit.docs.StateInfo.State
+import nl.knaw.dans.lib.logging.DebugEnhancedLogging
+import nl.knaw.dans.lib.error._
+
+import scala.util.Try
+
+object WorkerThread extends DebugEnhancedLogging {
+
+  val stagedBaseDir = File("target")
+
+  def main(args: Array[String]): Unit = {
+
+    // TODO create watchService as first action in EasyDepositApiService.start
+    //  close it as last action in EasyDepositApiService.stop
+    stagedBaseDir.watchService.apply { watchService =>
+      stagedBaseDir.register(watchService, Seq(ENTRY_CREATE, ENTRY_DELETE))
+      handleAbortedByHardShutdown() // between register and take of watchService!
+      while ( { // TODO in a sub-thread (possibly with handleAbortedByHardShutdown, for a quick start)
+        // https://howtodoinjava.com/java8/java-8-watchservice-api-tutorial/
+        // take, because polling skips staged-upload-dirs created and deleted within a single cycle
+        Try(watchService.take())
+          .map(handleEvents)
+          .doIfFailure { case e =>
+            // does not get here on SIGINT despite InterruptedException thrown by take
+            // TODO would it get a ClosedWatchServiceException when the parent thread is the culprit?
+            logger.error(s"terminating ${ e.getMessage }")
+          }
+          .getOrElse(false)
+      }) {}
+    }
+  }
+
+  private def handleEvents(watchKey: WatchKey): Boolean = {
+    watchKey.pollEvents.stream.forEach(handleEvent(_))
+    watchKey.reset() // next pending event
+  }
+
+  private def handleEvent(event: WatchEvent[_]): Unit = {
+    val stagedDir = stagedBaseDir / event.context().toString
+    (event.kind(), getDepositState(stagedDir.name)) match {
+      case (ENTRY_CREATE, State.finalizing) => finalizeSubmit(stagedDir)
+      case (ENTRY_DELETE, State.draft) => calculateShas(stagedDir)
+      case _ => // either completed finalizeSubmit or started upload
+    }
+  }
+
+  private def handleAbortedByHardShutdown(): Unit = {
+    stagedBaseDir.list.foreach { abortedDir =>
+      getDepositState(abortedDir.name) match {
+        case State.draft => abortedUpload(abortedDir)
+        case State.finalizing => abortedSubmit(abortedDir)
+        case state => logger.error(s"Staged garbage : $state $abortedDir")
+      }
+    }
+    logger.info(s"So far what was found at startup in $stagedBaseDir")
+  }
+
+  private def abortedSubmit(stagedSubmitDir: File): Unit = {
```

Comment:
```suggestion
  private def retrySubmit(stagedSubmitDir: File): Unit = {
```

Reply:
Both `aborted` methods start with logging `found ... aborted by hard shutdown`. They are called side-by-side from `handleAbortedByHardShutdown`. So I choose to call both methods `aborted`. But indeed one retries it, the other only triggers a retry to be executed via the `watchService.take` loop.